server/etcdserver/api/membership/cluster.go

```diff
@@ -524,26 +524,72 @@ func (c *RaftCluster) PromoteMember(id types.ID, shouldApplyV3 ShouldApplyV3) {
 		isLearner.Set(0)
 	}
 
-	if c.be != nil && shouldApplyV3 {
-		c.members[id].RaftAttributes.IsLearner = false
-		c.updateMembershipMetric(id, true)
-		c.be.MustSaveMemberToBackend(c.members[id])
+	if c.be != nil {
+		m := c.members[id]
+		if shouldApplyV3 {
+			m.RaftAttributes.IsLearner = false
+			c.updateMembershipMetric(id, true)
+			c.be.MustSaveMemberToBackend(m)
 
-		c.lg.Info(
-			"promote member",
-			zap.String("cluster-id", c.cid.String()),
-			zap.String("local-member-id", c.localID.String()),
-			zap.String("promoted-member-id", id.String()),
-		)
+			c.lg.Info(
+				"promote member",
+				zap.String("cluster-id", c.cid.String()),
+				zap.String("local-member-id", c.localID.String()),
+				zap.String("promoted-member-id", id.String()),
+			)
+		} else {
+			// Workaround the issues which have already been affected by
+			// https://github.com/etcd-io/etcd/issues/19557. The learner
+			// promotion request had been applied to v3store, but not saved
+			// to v2snapshot yet when in 3.5. Once upgrading to 3.6, the
+			// patch here ensure the issue can be automatically fixed.
+			if m.IsLearner {
+				m.RaftAttributes.IsLearner = false
+				c.lg.Info("Forcibly apply member promotion request", zap.String("member", fmt.Sprintf("%+v", *m)))
+				c.be.MustSaveMemberToBackend(m)
```

Comment:
MustSaveMemberToBackend will use `LockInsideApply` which will flush the CI. I don't remember what is state of CI when `shouldApplyV3`, but I assume it's pointing to the raftindex of the promote request. 

I think this is the place to call `MustHackySaveMemberToBackend`

Reply:
Conceptually, you are right, we should call `MustHackySaveMemberToBackend` here. But actually the result is the same. When `shouldApplyV3` is false, even calling `LockInsideApply` won't advance the CI.

Anyway, changed to `MustHackySaveMemberToBackend`, thx.

Reply:
Thanks for confirming and updating code.